### ISSUE TITLE
Create `processor` table

### ISSUE THREAD

**Issue by cgreene** [AlexsLemonade/refinebio#322]
### Context

We have been having the reproducibility and provenance meeting.

### Problem or idea

We would like to have a way to store the version information for a processor. We would like this to be populated automatically.

### Solution or next step

We should implement a processor table that is automatically populated. When a processor's docker image is started it will check this table for an existing row that matches it. If this is not found, it will extract the version information for "key dependencies" and store this in a JSON field. We should store the `primary program` (either as a field or in a table - if in a table we need to update it programmatically as the processor field is updated). As this conversation is ongoing we are moving a little bit away from a separate table, so let's go with the field route unless, on reflection, we come up with a strong reason to do something different.

### New Issue Checklist

- [x] The title is short and descriptive.
- [x] You have explained the context that led you to write this issue.
- [x] You have reported a problem or idea.
- [x] You have proposed a solution or next step.

**Comment by dongbohu** [comment]
Right now the `Processor` model is defined like this:
```
class Processor(models.Model):
    """Processor associated with a certain ComputationalResult."""

    name = models.CharField(max_length=255)
    docker_image = models.CharField(max_length=255)
    environment = JSONField(default={})

    class Meta:
        db_table = "processors"
        unique_together = ('name', 'docker_image')

    def __str__(self):
        return "Processor: %s (docker_image: %s" % (name, docker_image)
```

`Salmon` pipeline includes the following 4 processors:
- Salmon Quant
- Salmontools
- Tximport
- MultiQC

What should be included in each processor's `environment` field?

**Comment by cgreene** [comment]
This YAML file would be expected at a defined location with a defined set of categories (`ubuntu`, `R`, `python` initially).

**Comment by dongbohu** [comment]
@cgreene: Originally I was thinking of using JSON format, but as far as I nkow, JSON doesn't allow comment lines, which is very annoying. YAML looks good to me. By the way, since we will populate the YAML fields to the `environment` of each processor, does it make more sense to define the YAML file as an array of different processors? The example you listed above is a little confusing to me.

**Comment by cgreene** [comment]
I'm thinking that the YAML file for each processor would define what, exactly, we want versioned at the database level (as opposed to just "go check the docker image" which is what we offer for all other packages). Then we'd have to write some code that goes at the end of the spin up process (if I understand the conversation with @kurtwheeler and @Miserlou correctly) that extracts the version information via apt, pip? and R. That information would then get stored in the JSON field in the database and be made available to the front end for styling.

**Comment by dongbohu** [comment]
@cgreene Are you saying that we should have a YAML file for each processor? If yes, then your example makes sense.

**Comment by cgreene** [comment]
I think that there should be some readily parseable file within each processor that defines what package versions should be considered database-worthy (and consequently frontend worthy). Ideally this would be something that @jaclyn-taroni and other computational biologists could easily add to if they discovered later that we wanted to capture certain package versions and expose them via the UI (of course, they would only be captured for new versions of a processor going forward). I figured YAML was likely to have good parsers available.

**Comment by jaclyn-taroni** [comment]
Ignoring Ubuntu since it's somewhat implied, here are the packages/dependencies I think are important to keep track of for each of these processors:

#### Salmon quant

`Salmon`

#### Salmontools

`salmontools`

#### tximport

`R`: `tximport`, Bioconductor version, and possibly `readr`

#### MultiQC

`FastQC`, `Salmon`, `MultiQC`

#### Affymetrix SCAN

`array_express` should probably really be called `affymetrix`! 

`R`: `SCAN.UPC`, `oligo`, `affy`, `affyio`, `Biobase`, `AnnotationDbi`, `limma`, `sva`, Brainarray version, Bioconductor version

Ideally, we'd also supply platform design (`pd.*`) info but that's on a _per-platform_ basis. I imagine someone interested in this particular bit of information is more likely to be familiar with Docker and could go check the image.

#### Illumina SCAN

`R`: `oligo`, `limma`, Bioconductor, the `illumina` `.db` packages (there are only 8 of these total so more manageable than Affy!)

#### agilent_twocolor

_For now_  

`R`: `SCAN.UPC` & Bioconductor version

We'll almost certainly add more once we really support this platform and need to do identifier conversion

#### Smasher

`python`: `sklearn`, `pandas` 

#### transcriptome_index

`Salmon`

#### No Op

This is essentially the gene identifier conversion pipeline. This is less clear to me at the moment. @Miserlou - have you settled on an R/tidyverse vs. python/pandas implementation? 

Note also that the really important bits, from my point of view, are the packages that were used to generate the data used for the conversion (see #348) that are not part of the processor.

**Comment by dongbohu** [comment]
I created a new issue to keep track of the table population: #361.

**Comment by dongbohu** [comment]
Partially solved by #346, also overlaps with #361.
